Add ORCID OAuth2 provider

ml-evs opened the pull request:
This PR adds support for ORCID (https://orcid.org/) and its sandbox server (https://sandbox.orcid.org) via a blueprint init option `use_orcid_sandbox`.

It is slightly tricky to test this as OAuth configs at ORCID require human intervention at ORCID's end, and even then, the ORCID sandbox does not allow non-HTTPS callback URIs (like localhost). I was able to use this code to authorize via the ORCID sandbox by monkey-patching the authorized URL in the base blueprint to point at [Google's OAuth playground](https://developers.google.com/oauthplayground/)  using the `openid` scope (might be useful if this was easier to do).

Thanks for flask-dance!

singingwolfboy commented on flask_dance/contrib/orcid.py:
```diff
@@ -0,0 +1,93 @@
+        scope=scope,
+        base_url=base_url,
+        authorization_url="https://orcid.org/oauth/authorize",
+        token_url="https://orcid.org//oauth/token",
```
Did you intend for this to have two slashes before `oauth`?

singingwolfboy commented on tests/contrib/test_orcid.py:
```diff
@@ -0,0 +1,101 @@
+    assert orcid_bp.token_url == "https://orcid.org/oauth2/token"
+
+
+def test_blueprint_factory():
```
```suggestion
def test_blueprint_factory_for_sandbox():
```

singingwolfboy reviewed (changes requested):
This is a great start; thanks for making this pull request! However, your tests are failing, and you still need to edit the [`docs/providers.rst`](https://github.com/singingwolfboy/flask-dance/blob/main/docs/providers.rst) file so that this shows up in the documentation. You also need to update the changelog. See the [README in the `contrib` dir](https://github.com/singingwolfboy/flask-dance/blob/main/flask_dance/contrib/README.rst).

ml-evs commented:
Thanks for the review, rushed this out a bit and should have left it as a draft until I could finish it off. I should have addressed all your concerns (apologies if you had to manually authorize rerun the workflows!)

singingwolfboy commented on flask_dance/contrib/orcid.py:
```diff
@@ -0,0 +1,97 @@
+    application config defines them, using the variables
+    :envvar:`ORCID_OAUTH_CLIENT_ID` and :envvar:`ORCID_OAUTH_CLIENT_SECRET`.
+
+    The ORCID Sandbox API (https://sandbox.orcid.org) will be used if
+    the ``use_orcid_sandbox`` argument is set to true.
```
I actually think that `sandbox` is a clear enough name. The function is already named `make_orcid_blueprint()`, including "orcid" in the argument name seems unnecessary. Let's keep it concise!

ml-evs replied:
Fair, it felt a bit ambiguous to me (given that its really just contacting a different OAuth server) but in context it should be clear enough!

singingwolfboy commented:
Looks great, thank you! I'll cut a new release in the next few days.

ml-evs commented:
> Looks great, thank you! I'll cut a new release in the next few days.

Thanks, I appreciate it!

singingwolfboy commented:
@ml-evs I just published version 6.2.0 a few hours ago, including your change! Thanks again for contributing to this project. 😄

ml-evs commented:
Awesome, thanks for letting me know.

mpound commented:
@ml-evs Do you have a working example using orcid for authentication you could share?  similar to the github example on https://flask-dance.readthedocs.io/en/latest/ ? 
I tried the obvious substitutions of 'orcid' for 'github' in the python script, but get weird errors like "Incorrect OAuth Link for client id , missing scope parameter."  However changing the call to explicitily set the scope: make_orcid_blueprint(scope=["authenticate"])   results in a different error "CSRF Warning! State not equal in request and response."     A working example might help clarify what I am doing wrong.
thanks.

ml-evs commented:
Hi @mpound, sure thing, I can type something up when I get home. Have you registered an app with the sandbox ORCID API already? For me it was process with manual approval from the ORCID team. 

Just in case it is a simple issue, I can see that in my code (unfortunately closed source for now) my scope is set to `"/authenticate"` rather than the list `["authenticate"]`.

mpound commented:
ok, I got a little farther with your hints.   It appears to authenticate correctly at
    resp = orcid.get("/user")
    assert resp.ok

But resp.ok is False.  I did register the test app with sandbox orcid on the developer tools section of my account, which gave me the CLIENT_ID and CLIENT_SECRET.
